ambry-network/src/test/java/com.github.ambry.network/SSLSelectorTest.java

```diff
@@ -206,33 +218,160 @@ public void testSSLConnect() throws IOException {
 
   @Test
   public void testCloseAfterConnectCall() throws IOException {
-    String connectionId =
-        selector.connect(new InetSocketAddress("localhost", server.port), BUFFER_SIZE, BUFFER_SIZE, PortType.SSL);
+    String connectionId = selector.connect(new InetSocketAddress("localhost", server.port), DEFAULT_SOCKET_BUF_SIZE,
+        DEFAULT_SOCKET_BUF_SIZE, PortType.SSL);
     selector.close(connectionId);
     selector.poll(0);
     Assert.assertTrue("Channel should have been added to disconnected list",
         selector.disconnected().contains(connectionId));
   }
 
+  /**
+   * selector.poll() should be able to fetch more data than netReadBuffer from the socket.
+   */
+  @Test
+  public void testSelectorPollReadSize() throws Exception {
+    for (int socketBufSize : new int[]{applicationBufferSize, applicationBufferSize * 3, applicationBufferSize - 10}) {
+      String connectionId = blockingSSLConnect(socketBufSize);
+      String message =
+          SelectorTest.randomString(socketBufSize * 2 + TestUtils.RANDOM.nextInt(socketBufSize), TestUtils.RANDOM);
+
+      // Did not count the exact number of polls to completion since its hard to make that test deterministic.
+      // i.e. EchoServer could respond slower than expected.
+      assertEquals("Wrong echoed response", message, blockingRequest(connectionId, message));
+    }
+  }
+
+  /**
+   * Tests handling of BUFFER_UNDERFLOW during unwrap when network read buffer is smaller than SSL session packet buffer
+   * size.
+   */
+  @Test
+  public void testNetReadBufferResize() throws Exception {
+    useCustomBufferSizeSelector(10, null, null);
+    String connectionId = blockingSSLConnect(DEFAULT_SOCKET_BUF_SIZE);
+    String message = SelectorTest.randomString(5 * applicationBufferSize, new Random());
+    assertEquals("Wrong echoed response", message, blockingRequest(connectionId, message));
+  }
+
+  /**
+   * Tests handling of BUFFER_OVERFLOW during wrap when network write buffer is smaller than SSL session packet buffer
+   * size.
+   */
+  @Test
+  public void testNetWriteBufferResize() throws Exception {
+    useCustomBufferSizeSelector(null, 10, null);
+    String connectionId = blockingSSLConnect(DEFAULT_SOCKET_BUF_SIZE);
+    String message = SelectorTest.randomString(10, new Random());
```

Comment:
I though `BUFFER_OEVERFLOW` is in `SSLTransmission::write(ByteBuffer)`, but it turns out no. In step debugging, I found `BUFFER_OVERFLOW` happened in `SSLTransmission::handshake()` only and `netWriteBuffer` is expanded to `16KB` after handshake. 

I also changed message size to 4*16KB to make it > `netWriteBuffer `, but `write(ByteBuffer)->warp()` still returns OK.  

No change needed, just FYI.

Reply:
Yep, unfortunately this test only tests the buffer expansion flow within the handshake code. If I think of a good way to test it within the write call, I'll publish a follow up PR.